ktor-features/ktor-metrics-micrometer/jvm/src/io/ktor/metrics/micrometer/MicrometerMetrics.kt

```diff
@@ -0,0 +1,185 @@
+ *       histogram is exposed. Use the [Configuration.timerCustomizer] to enable these.
+ *       Tags:
+ *       <ul>
+ *           <li><code>local</code>: The host and port of the request uri (e.g. 'www.ktor.io')</li>
+ *           <li><code>method</code>: The http method (e.g. 'GET')</li>
+ *           <li><code>route</code>: The use ktor route used for this request. (e.g. '/some/path/{someParameter}')
+ *           <li><code>status</code>: The http status code or the Exception name thrown during the request (e.g. '200', 'java.lang.IllegalArgumentException')
+ *        <ul>
+ *     <li>
+ *  <ul>
+ */
+class MicrometerMetrics(
+    private val registry: MeterRegistry,
+    timerDistributionConfig: DistributionStatisticConfig,
+    private val tagCustomizer: MicrometerMetrics.TagBuilder.(ApplicationCall) -> Unit
+) {
+
+    private val active = registry.gauge("$baseName.requests.active", AtomicInteger(0))
+
+    init {
+        enableTimerDistributionConfig(timerDistributionConfig)
+    }
+
+    private fun enableTimerDistributionConfig(timerDistributionConfig: DistributionStatisticConfig) {
+        registry.config().meterFilter(object : MeterFilter {
+            override fun configure(id: Meter.Id, config: DistributionStatisticConfig): DistributionStatisticConfig =
+                if (id.name == requestTimerName)
+                    timerDistributionConfig.merge(config)
+                else
+                    config
+        })
+    }
+
+    /**
+     * Configures this Feature
+     * @property registry The meter registry where the meters are registered. Mandatory
+     * @property meterBinders The binders that are automatically bound to the registry. Default: [ClassLoaderMetrics],
+     * [JvmMemoryMetrics], [ProcessorMetrics], [JvmGcMetrics], [ProcessorMetrics], [JvmThreadMetrics], [FileDescriptorMetrics]
+     * @property distributionStatisticConfig configures the histogram and/or percentiles for all request timers. By
+     * default 50%, 90% , 95% and 99% percentiles are configured
+     * @property tagBuilder is called to customize the tags of the timer for each call. by default this is a no-op
+     */
+    class Configuration {
+
+        lateinit var registry: MeterRegistry
+
+        var meterBinders: List<MeterBinder> = listOf(
+            ClassLoaderMetrics(),
+            JvmMemoryMetrics(),
+            JvmGcMetrics(),
+            ProcessorMetrics(),
+            JvmThreadMetrics(),
+            FileDescriptorMetrics()
+        )
+
+        var distributionStatisticConfig: DistributionStatisticConfig =
+            DistributionStatisticConfig.Builder()
+                .percentiles(0.5, 0.9, 0.95, 0.99)
+                .build()
+
+        var tagCustomizer: TagBuilder.(ApplicationCall) -> Unit = {}
+    }
+
+    private fun ApplicationCall.getLocalAddress() = request.local.let { "${it.host}:${it.port}" }
+
+    private fun ApplicationCall.getStatusCode() = response.status()?.value?.toString() ?: "Unknown"
```

Comment:
I changed the generation for this value, now it should be clear that the "unknown" case should not happen as there is either a `Throwable` or the status is set. 

@cy6erGn0m can it happen that both (throwable and status) are set?

Reply:
It could happen if no handler matched. So I'd say if we have no status and no exception then it is 404.

Reply:
@JotBePunkt this case still may cause failure `"Sorry, this should not happen, we should have either a  http status code or a throwable"`

Reply:
I wrote some tests and in fact all four cases can happen:
- No Status, No Exception (no handler registered) Here it might be good, if the ktor framework would either generate a 404, log an error message or something else. This should not be droppen silently) 
- Status Set: No exception (normal case)
- No status Set, Exception: This happens when the exception is thrown before the status is set (in this case ktor might render a Stautus 500 for the client)
- Status set: Exception: This happen when the exception is thrown after the status is set. I tested iit only with the test framework and not with a real http client but this might also be strage for the client, when an exception is thrown bit the clients stiill gets Status 200.

Reply:
Update: I found the point where the default status codes are applied:  `io.ktor.server.engine.DefaultEnginePipelineKt.defaultEnginePipeline` unfortunately this pipeline cannot be accessed from an Application feature but we can intercept the SendPipeline. That is just a but hard to test as the TestEngine replaces this behaviour with a test behaviour.

Reply:
Finally... had do change the TestEngine so we can change be behaviour to emulate the default error handling also in tests (e.g respond status 404 when no handler resonded). Due to the fact that the Timer is now stopped in the sendPipeline (which is executed when the method `respond(...)` is called) the Metrics feature cannot track exceptions that are thrown *after* the `respond()` method. It would be great if we would could get all exceptions but since usually the `respond(...)` is usually the last method called, this should be ok.